enhancements/20250522-expose-cdi-settings-in-ui.md

```diff
@@ -0,0 +1,114 @@
+
+**After**: User can directly modify StorageProfile-related settings through the Harvester UI, using intuitive form controls and dropdowns with clear explanations. All changes are automatically applied to the underlying StorageProfile CRDs.
+
+### User Experience In Detail
+
+The CDI settings will be integrated into the existing Storage Classes page in Harvester UI.
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### UI Preview
+
+![storageclass-ui](./20250522-expose-cdi-settings-in-ui/storageclass-ui.png)
+
+- Note: These new tab: CDI Settings is not available in the longhorn-v1 StorageClass.
+- The new CDI settings can be configured during the creation of a StorageClass or by updating a StorageClass.
+- The Volume Mode / Access Modes setting supports multiple combinations, where each combination must have a unique Volume Mode and one or more associated Access Modes.
+  - The volume mode should be either `FileSystem` or `Block`.
+  - The access modes can have 0~4 unique mode, the access mode could be either `ReadWriteOnce`, `ReadOnlyMany`, `ReadWriteMany` or `ReadWriteOncePod`.
+- The Volume Snapshot Class is presented as a dropdown menu listing all available VolumeSnapshotClass names.
+- The Clone Strategy is also a dropdown menu with the following fixed options: "copy", "snapshot", and "csi-clone".
+- The File System Overhead must be a value between 0 and 1, with a regex pattern of ^(0(?:\.\d{1,3})?|1)$ (i.e., up to three digits after the decimal point).
+
+#### Backend Implementation
+
+Introduce 4 annotations to StorageClass:
+
+- `cdi.harvesterhci.io/filesystemOverhead`: String value representing filesystem overhead percentage (0-1), regex pattern: `^(0(?:\.\d{1,3})?|1)$` (up to 3 digits after the decimal point)
+- `cdi.harvesterhci.io/cloneStrategy`: String value, one of: "copy", "snapshot", "csi-clone"
+- `cdi.harvesterhci.io/volumeSnapshotClass`: String value, must be an existing VolumeSnapshotClass name
+- `cdi.harvesterhci.io/volumeModeAccessModes`: JSON string mapping VolumeMode to AccessMode arrays
+
+Below is the StorageClass yaml example:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+    cdi.harvesterhci.io/filesystemOverhead: "0.85"
+    cdi.harvesterhci.io/cloneStrategy: "snapshot"
+    cdi.harvesterhci.io/volumeSnapshotClass: "csi-snapshot-class"
+    cdi.harvesterhci.io/volumeModeAccessModes: |
+      {"Block":["ReadWriteOnce","ReadOnlyMany"],"Filesystem":["ReadWriteOnce","ReadWriteMany"]}
+```
+
+The implementation consists of three main components:
+
+1. StorageClass Webhook Validation
+Extend the existing StorageClass validation webhook to validate the five new annotations. The webhook will perform pattern matching for filesystemOverhead, enum validation for cloneStrategy, existence checks for volumeSnapshotClass, and JSON schema validation for volumeModeAccessModes.
```

Comment:
nit: Does CDI have the corresponding validating admission? If so, we might be able to delegate the validation to it, as I assume it's validated synchronously during configuration. The downside could be that the error message can't be customized.

Reply:
>Therefore, we still need to implement our own validation logic for these annotations defined in StorageClass.
minor: I see, but the validation will be the same as what CDI does. Not sure if we can use their validation using CDI as a lib to prevent duplicated implementation or future change in validation.